### ISSUE TITLE
Mixing program-wide and command-specific options may fail depending on their order

### ISSUE THREAD

**Issue by lpotmann** [pcapriotti/optparse-applicative#23]
```
import Options.Applicative
import Control.Applicative

data Program = Program { progOption :: String
                       , progCommand :: Command
                       }
  deriving (Show)

data Command = Command { commandOption :: String }
  deriving (Show)

commandOpts :: Parser Command
commandOpts = Command <$> strOption ( long "command-option" )

options :: Parser Program
options = pure Program
  <*> strOption ( long "program-option" )
  <*> subparser ( command "command" ( info commandOpts idm ) )

main :: IO ()
main = execParser (info options idm) >>= print

```

Example of the order-specific behaviour:

```
$ runghc Test.hs command --program-option=foo --command-option=bar
Usage: Test.hs command --command-option ARG
$ runghc Test.hs command --command-option=bar --program-option=foo
Program {progOption = "foo", progCommand = Command {commandOption = "bar"}}
```

I would expect both of these to work, given that there's no ambiguity in the option names.

**Comment by ababkin** [comment]
I actually strongly disagree and I think the issue should be re-opened

I have groups of commands and have various options that are specific to commands/groups.
Because the Commands are nested into Groups, I get this understandable current behavior (because of the way the `subparser` works), but it feels very weird to my users.
My every Command has a set of unique options in addition to a set of options that are the same for all commands in a particular Group. If I parse the unique options into the `Command` data type (in a `subparser`) and the common group options into the `Group` data type, the order of options starts to matter and is jarring to my users. (I understand the whole point of `optparse` being based on `Applicative` is that the order of options **should not** matter)

Because of the current behavior, to make order not matter, I'm stuck with duplicating the common fields that get parsed from the Group options into every Command (which is clearly suboptimal), i.e. I dont parse any options on the Group level.

I hope my description makes sense. I can provide a concrete example if it helps to illustrate my point

Thanks
Alex

**Comment by HuwCampbell** [comment]
I'm a bit adverse to reopening 5 year old issues, but will admit that I have found this unsatisfactory as well. In my (incomplete) idris port I actually implemented subparsers in a different way to avoid this.

For now, have a look at and try my [subparser-reimplementation-breaking](https://github.com/pcapriotti/optparse-applicative/compare/topic/subparser-reimplementation-breaking) branch which uses those ideas. It changes how subparsers work, by instead of executing them and then inserting the value into the top level parser, just substituting the chosen parser branch into the parser as is.

I haven't merged it because it regresses a handful of tests and completely removes the no-backtracking facility that a small number of programs rely on.

If that branch solves your problem though, I can reach out to for a wider comment.
